### DISCUSSION TITLE
PDF to SVG looks weird (like hand drawn)

### DISCUSSION THREAD

**Discussion by paigenorth** [ImageMagick/ImageMagick#2457] [Help] [answered]
Running `magick test.pdf test.svg` I find that [test.pdf](https://github.com/ImageMagick/ImageMagick/files/5114285/test.pdf) is converted to:
![test](https://user-images.githubusercontent.com/7458712/90983306-1d8f8e80-e53b-11ea-97b7-864826bcfe16.png) 
I've looked through the documentation, but I can't find what might be causing this strange `hand-drawn' effect. 

I am using ImageMagick 7.0.10-28 on macOS Catalina 10.15.3.

Any help would be greatly appreciated!

**Comment by snibgo** [comment]
But you don't show us the SVG. Instead, you show us a PNG file. What converted the SVG to the PNG?

**Reply by paigenorth** [reply]
Github won't allow me to attach an SVG file, so I converted it to PNG using ImageMagick. It looks exactly the same, but if it can be useful, here is the original SVG file zipped up:
[test_magick.svg.zip](https://github.com/ImageMagick/ImageMagick/files/5114859/test_magick.svg.zip)

**Comment by snibgo** [comment, marked as answer]
The PDF has vector data. Your SVG has been created by potrace. So IM has called Ghostcript to raterize the PDF vector, then called potrace to convert it back to vector, and stored those vectors in a SVG file. Then you rasterize it again.

The result looks better if we don't use potrace, and store the rasterised image in the SVG.

What do you really want to do? If you want to convert the PDF vectors into SVG vectors, ImageMagick can't do that. Perhaps some other program can.

**Reply by paigenorth** [reply]
I didn’t realize that’s what ImageMagick was doing. Thanks for the info. I’ll look for a more direct way to convert pdf to svg.

**Comment by tillsanders** [comment]
For anyone else coming here: I ultimately used Poppler / [pdftocairo](https://docs.oracle.com/cd/E88353_01/html/E37839/pdftocairo-1.html) for this specific task.

**Comment by hunter1749** [comment]
Something isn't adding up. I get that @snibgo already explained what the command does, but I get a rasterised output no matter the PDF I feed in, which is to be expected from ImageMagick. I'm on Windows, so could it be that it works differently on macOS? And if @paigenorth managed to _successfully_ convert PDF to SVG, what's stopping ImageMagick from doing the reverse (SVG to PDF)?

**Reply by fmw42** [reply]
To go from SVG to PDF, you need an SVG reader/rederer. Imagemagick can use any one of 3 different ones: MSVG/XML (imagemagick's internal default renderer), RSVG delegate or Inkscape (in order of increasing quality/relieability).  Which one do you have?  Recommend that you install Inkscape. Imagemagick will use that automatically over any of the other 3 without having to re-install Imagemagick.

**Reply by hunter1749** [reply]
Well I have at least Inkscape installed, but does that effect the conversion from PDF to SVG? ImageMagick isn't made to convert between vector formats, so I'm confused as to how OP got an actual SVG full of different strokes as a result, whereas I always just get an image embedded into the skeleton of SVG. It's as though Potrace is never called by my ImageMagick after it's done with the rasterising step. I've included the output and input side by side. Apart from this post, I couldn't find anyone else trying to force a true PDF to SVG conversion in ImageMagick. That's what got me commenting here, because I was sure this conversion would always produce some rasterised output. This post proved me wrong it seems.
![image](https://github.com/user-attachments/assets/a9026bcc-944d-4e02-a7c1-638d3f8e380a)
Edit: In case it's not clear, on the left is very pixelated SVG, and on the right is a pure PDF.

**Comment by fmw42** [comment]
No to go from PDF to SVG requires that you also have Ghostscript to read the PDF.

**Reply by hunter1749** [reply]
I also have Ghostscript working.

**Comment by snibgo** [comment]
> I'm confused as to how OP got an actual SVG full of different strokes...

As I said above, the OP used `potrace` to read a raster image and create vector SVG.

**Reply by hunter1749** [reply]
The way you wrote it then, it seemed ImageMagick utilised Potrace by itself. So you do agree that the command `magick test.pdf test.svg` alone doesn't produce _an editable_ SVG, but instead an image embedded into an SVG? For the SVG to become _editable_, at least two commands would be required (one for `PDF --ImageMagick--> raster image`, and another for `raster image --Potrace--> SVG`)?

**Comment by snibgo** [comment]
That was four years ago, and I can't remember why I said that IM "then called potrace". Perhaps it did, or perhaps the user did this as a separate command. If you look at the posted SVG file, it says it was created by potrace.

IM can use potrace as an external delegate, by putting an entry in `delegates.xml`, such as:
```
  <delegate decode="pnm" encode="svgv" command="&quot;potrace&quot; --svg -o &quot;%o&quot; &quot;%i&quot;"/>
```
Then we can vectorize a raster image like this:
```
magick in.png svgv:out.svg
```
If we omit the prefix `svgv:`, the image will not be vectorized, and the created svg will contain an embedded raster image.

EDIT: I should stress that when IM reads a PDF file, it uses Ghostscript to rasterize the PDF pages. So IM then has one or more raster images. IM can then use potrace to vectorize the image(s). IM can't read PDF vectors and directly create the corresponding SVG vectors.

**Comment by hunter1749** [comment]
> That was four years ago, and I can't remember why I said that IM "then called potrace". Perhaps it did, or perhaps the user did this as a separate command. If you look at the posted SVG file, it says it was created by potrace.

That explains it for me. So, as of making this reply, the command `magick test.pdf test.svg` leaves the converted SVG rasterised in contrast to the result shown from 2020 where it was additionally processed by Potrace.

And just to clarify:
* The original post's SVG is made by Potrace, as seen in its contents.
* Somehow, only the command `magick test.pdf test.svg` was needed to create that SVG.
* But achieve the same result today, you would need to:
  * Either first convert the PDF into a raster format using ImageMagick, and then convert the raster format into SVG using Potrace.
  * Or use the external delegate method you described.
